pinot-common/src/main/java/org/apache/pinot/common/datablock/MetadataBlock.java

```diff
@@ -36,108 +39,129 @@
-  @VisibleForTesting
-  static class Contents {
-
-    private String _type;
-    private Map<String, String> _stats;
-
-    @JsonCreator
-    public Contents(@JsonProperty("type") String type, @JsonProperty("stats") Map<String, String> stats) {
-      _type = type;
-      _stats = stats;
-    }
-
-    @JsonCreator
-    public Contents() {
-      this(null, new HashMap<>());
-    }
-
-    public String getType() {
-      return _type;
-    }
+  public MetadataBlock(MetadataBlockType type, List<ByteBuffer> statsByStage) {
+    super(0, null, new String[0], new byte[]{(byte) (type.ordinal() & 0xFF)}, new byte[0]);
+    _type = type;
+    _statsByStage = statsByStage;
+  }
 
-    public void setType(String type) {
-      _type = type;
+  MetadataBlock(ByteBuffer byteBuffer)
+      throws IOException {
+    super(byteBuffer);
+    // Remember: At this point deserializeMetadata is already being called.
+    if (_fixedSizeDataBytes == null) {
+       if (_errCodeToExceptionMap.isEmpty()) {
+         _type = MetadataBlockType.EOS;
+       } else {
+         _type = MetadataBlockType.ERROR;
+       }
+    } else {
+      _type = MetadataBlockType.values()[_fixedSizeDataBytes[0]];
     }
+  }
 
-    public Map<String, String> getStats() {
-      return _stats;
+  @Override
+  protected void serializeMetadata(DataOutputStream output)
+      throws IOException {
+    if (_statsByStage == null) {
+      output.writeInt(0);
+      return;
     }
-
-    public void setStats(Map<String, String> stats) {
-      _stats = stats;
+    int size = _statsByStage.size();
+    output.writeInt(size);
+    if (size > 0) {
+      byte[] bytes = new byte[4096];
+      for (ByteBuffer stat : _statsByStage) {
+        if (stat == null) {
+          output.writeBoolean(false);
+        } else {
+          output.writeBoolean(true);
+          output.writeInt(stat.remaining());
+          ByteBuffer duplicate = stat.duplicate();
+          while (duplicate.hasRemaining()) {
+            int length = Math.min(duplicate.remaining(), bytes.length);
+            duplicate.get(bytes, 0, length);
+            output.write(bytes, 0, length);
+          }
+        }
+      }
     }
   }
 
-  private final Contents _contents;
-
-  public MetadataBlock(MetadataBlockType type) {
-    this(type, new HashMap<>());
-  }
-
-  public MetadataBlock(MetadataBlockType type, Map<String, String> stats) {
-    super(0, null, new String[0], new byte[]{0}, toContents(new Contents(type.name(), stats)));
-    _contents = new Contents(type.name(), stats);
-  }
-
-  private static byte[] toContents(Contents type) {
-    try {
-      return JSON.writeValueAsBytes(type);
-    } catch (JsonProcessingException e) {
-      throw new RuntimeException(e);
+  public static MetadataBlock deserialize(ByteBuffer byteBuffer, int version)
+      throws IOException {
+    switch (version) {
+      case 1: {
+        V1MetadataBlock decoded = new V1MetadataBlock(byteBuffer);
```

Comment:
I'm still confused. Is this the only place where v1 metadata block is used? Does it handle error differently from v2 metadata block?
Does it work if we ignore the version and always `return new MetadataBlock(byteBuffer);`?
If the error handling is different, then sending v2 metadata block to old server will cause backward incompatibility right?

Reply:
>  Is this the only place where v1 metadata block is used?
Yes. It is used only here and in the tests.

> Does it work if we ignore the version and always return new MetadataBlock(byteBuffer);?
Yes. 

> If the error handling is different, then sending v2 metadata block to old server will cause backward incompatibility right?

It is no different.

Remember that the main reason to have this class is to be able to run tests with it. The fact that it is useful here is just an extra.

Reply:
I planned to move V1MetadataBlock to the test classes, but I found some issues and I think the correct thing is to keep them here.

It looks like we cannot always return `new Metadata(byteBuffer)` because errors are encoded in a different way. Specifically, the metadata type (whether it is a eos or an error) was encoded in the variable part in V1 and in the fixed part in V2.

The change is still forward compatible (V1 can read V2) and with these lines, V2 can also read V1. I've created more test to verify that.

Reply:
Hmm, if we are encoding errors differently in v1 and v2, how does v1 understand the errors in v2?

Reply:
Let's list down the behavior on reading v1 bytes as v2 and also reading v2 bytes as v1 so that user knows what to expect when upgrading.
With current way of keeping both versions, we don't need to read v1 bytes as v2, but we can still read v2 bytes as v1 block

Reply:
> Hmm, if we are encoding errors differently in v1 and v2, how does v1 understand the errors in v2?

In V1 the quality of being an error is reported in two different ways:
1. By having a map of exceptions, which is serialized in a specific byte buffer section.
2. By setting the `type` in the json `Contents` stored in the variable size section.

In V2 we just use the first option. I that is what the very old _V0_ (whose source code is in `MetadataBlockTest.V0MetadataBlock`) in version was doing. Given V1 can read V0, it works as expected.

Specifically, V1 is able to read bytes where the variable section is empty or `Contents.type` is not defined because `getType` is defined as:

```java
  public MetadataBlock.MetadataBlockType getType() {
    String type = _contents.getType();

    // if type is null, then we're reading a legacy block where we didn't encode any
    // data. assume that it is an EOS block if there's no exceptions and an ERROR block
    // otherwise
    return type == null
        ? (getExceptions().isEmpty() ? MetadataBlock.MetadataBlockType.EOS : MetadataBlock.MetadataBlockType.ERROR)
        : MetadataBlock.MetadataBlockType.valueOf(type);
  }
```

Given we don't set `Contents.getType`, its value is null and we return EOS or ERROR depending on whether exceptions is empty or not.

Reply:
> Let's list down the behavior on reading v1 bytes as v2 and also reading v2 bytes as v1 so that user knows what to expect when upgrading.

The compatibility matrix is verified by tests. I've just renamed the name of the tests and I'm adding here the test table to show which tests verifies each scenario:

| input version | input type | input content | possible? | read in version | output type | output content |                   tested by                   |
|:-------------:|:----------:|:-------------:|:---------:|:---------------:|:-----------:|:--------------:|:---------------------------------------------:|
|      v1       |    eos     |     empty     |    yes    |       v2        |     eos     |     empty      |  v1EosWithStatsIsDecodedAsV2EosWithoutStats   |
|      v1       |    eos     |   not empty   |    yes    |       v2        |     eos     |     empty      | v1EosWithoutStatsIsDecodedAsV2EosWithoutStats |
|      v1       |   error    |     empty     |    no     |        -        |      -      |       -        |                       -                       |
|      v1       |   error    |   not empty   |    yes    |       v2        |    error    | same as input  |  v1ErrorIsDecodedAsV2ErrorWithSameExceptions  |
|      v2       |    eos     |     empty     |    yes    |       v1        |     eos     |     empty      | v2EosWithoutStatsIsReadInV1AsEosWithoutStats  |
|      v2       |    eos     |   not empty   |    yes    |       v1        |     eos     |     empty      |   v2EosWithStatsIsReadInV1AsEosWithoutStats   |
|      v2       |   error    |     empty     |    no     |        -        |      -      |       -        |                       -                       |
|      v2       |   error    |   not empty   |    yes    |       v1        |    error    | same as input  |  v2ErrorIsReadInV1AsErrorWithSameExceptions   |

Reply:
Given the above chart, do we need to have the special handling here?
This is a non-blocking comment, we may iterate on this

Reply:
If we remove this case, then `v1ErrorWithExceptionsIsDecodedAsV2ErrorWithSameExceptions` fails because we decode as EOS instead of error.

The reason for that is because when I modified `MetadataBlock` I decided to keep the `_type` attribute. But I guess it doesn't make much sense if we define that an error block must always contain at least one exception (like in _V0_). I've just changed that and now all tests are passing and V1MetadataBlock can be moved to the test directory.

Anyway, one of my plans is to clean this code a lot in order to improve serialization/deserialization times.